### PR TITLE
Add interchange data model description + JSON Schema definition

### PR DESCRIPTION
**Pull request by eemeli**
This proposes a way for us to fulfill our [deliverable](https://github.com/unicode-org/message-format-wg/blob/main/docs/goals.md) of:
> A formal definition of the canonical data model for representing localizable _dynamic message strings_.

This is not a data model that implementations MUST use, but one which they MAY support. It's intended to work as a formalization of the _meaning_ of our syntax as well as an interchange format for messages in other syntaxes, such as MF1. Effectively, it's synonymous with our syntax, but expressed as a parsed JSON structure.

It is intended to be a format capable of representing messages in all current syntaxes. As in, you should be able to parse anything into this structure with no data loss, and then use a compatible MF2 library with your messages. Full roundtripping of formats that support inline selectors like MF1 and Fluent might not retain their original structure, but will be semantically equivalent.

The format is included inline in the markdown doc using TypeScript interfaces, but its canonical form is provided as a [JSON Schema](https://json-schema.org/) definition.

### CODE REVIEW

**Comment by cdaringe** [comment]
1. Could this PR be considered part of an _unmentioned_ step in https://github.com/unicode-org/message-format-wg/pull/396? Could this data-model is may be just one of many in play? This model reads to me like a `ParsedMessage` data-model, where message parsing is actually the first unmentioned step of #396.
2. I can forsee other possible data models, like your `MessageValue` structure, also being a spec deliverable.

In the case of a central message provider dispatching messages to N-type-clients, perhaps in some systems:

- sending the raw `ParsedMessage` model over-the-wire could be applicable (data was not available to the central provider), or
- sending the `MessageValue` over the wire could be applicable (data was available to the central provider),

...depending on data availability in either context. Being able to distribute the parsing _could_ be beneficial (as clearly evidenced by this PR)!

**Comment by eemeli** [comment]
> This model reads to me like a `ParsedMessage` data-model, where message parsing is actually the first unmentioned step of #396.

That's a good point; the formatting spec should really mention that it's getting a parsed message structure from _somewhere_ to start with. We should not require parsing syntax into the data model specified here before formatting, but the synonymity between the data model and the syntax should allow for messages from either source (syntax or data model) to be equivalently formatted.

> I can forsee other possible data models, like your `MessageValue` structure, also being a spec deliverable.

That is a possibility, but tbh it's somewhat unlikely. For `Intl.MessageFormat` the `MessageValue` structure makes sense, as there it's intended to act as a base platform for further wrapping formatters. Consider a DOM/HTML message formatter: For something like that, it makes much more sense to output DOM fragments rather than MessageValues. Similarly in other settings, the surrounding environment will strongly guide the shape of non-string formatting targets.

I don't think there is sufficient value in harmonising the output format of MF2 to overcome the costs that would be incurred by implementations needing to fit its mold.

> sending the raw `ParsedMessage` model [...] or sending the `MessageValue` over the wire could be applicable

Parsing something like [MessageResource](https://github.com/eemeli/message-resource-wg/) _should_ be comparable to parsing JSON, and formatting a message _should_ be a rather fast operation, so I rather hope that we'll be able to not need such optimizations. But yes, there is value in portable representations of messages.

I see the following as probably the greatest benefits provided by this data model:

1. **Easing the initial or partial adoption of MF2.** Consider an existing application or other system that already has messages in some legacy format, tooling and pipelines for their translation, and a runtime for their formatting.

   With a well defined data model, it becomes easier for a step-by-step transition to MF2 to take place. Initially, the runtime parser could be replaced with a parser targeting the MF2 data model and and the formatter with an MF2 implementation. Next, the message data could be transformed to MF2 during the application build. Further steps could replace the whole localization stack with MF2.

   Without a data model, it becomes much harder to progress in such small steps, and the immediate cost of starting to use MF2 becomes much higher.

2. **Message transformations.** Not only between message formats, but also when message contents need to change. With a data model, it's much easier for interoperable tooling to be able to e.g. update a variable reference in all localized versions of a message.

3. **Providing a universal "message" definition.** This one is admittedly a bit fuzzier than the preceding, but potentially even more valuable. As far as I know, this data model is the first one that even attempts to be compatible with _all_ message formats. With this, you should be able to take a message in any existing syntax and represent it as structured data. I don't really know what people might do with it, but providing a qualitatively new tool is bound to be useful in all sorts of ways.

File: spec/message.json
Comment (catamorphism) [inline spec/message.json:131]: Seems like it might be simpler to have a single `message`, which has a `body` attribute that's one of `pattern` or `select`, and then `select` would just have the `selectors` and `variants` fields? (In other words, `declarations` would be factored out.)

File: spec/data-model.md
Comment (eemeli) [reply spec/data-model.md:16]: I don't think we'll be changing the structure much any more, and I'd say that the experience with the ABNF would indicate that the need to update two different places when applying changes is not too onerous. And yes, the explicit reference to the JSON Schema as authoritative is intended to disambiguate any divergences, should such form.

File: spec/data-model.md
Comment (eemeli) [reply spec/data-model.md:46]: Good point; calling this a "ref" is a bit wrong. Using "name" either just here or throughout could work.

One option would be to use something like this, much like `Option`.
```suggestion
  name: string
```

I don't think there's any parsing metadata that couldn't be encoded within `Declaration`.

File: spec/data-model.md
Comment (eemeli) [reply spec/data-model.md:52]: No, you've understood right. That is a theoretical possibility, but I think it's either purely theoretical, or represents actually different catchall key values.

The only existing syntax in which I could imagine this situation arising is Fluent, with a syntactical structure like this:
```fluent
key =
  { $one ->
     [1]
       { $two ->
         *[foo] Foo
       }
    *[2]
       { $two ->
         *[bar] Foo
       }
  }
```

With that, when you merge the inline selectors to a top-level selector, you could disambiguate the two `$two` selectors into one while keeping their default case values different, leading to an MF2 message like this:
```js
{
  type: 'select',
  declarations: [],
  selectors: [
    { type: 'variable', name: 'one' },
    { type: 'variable', name: 'two' }
  ],
  variants: [
    {
      keys: [
        { type: 'literal', quoted: false, value: '1' },
        { type: '*', value: 'foo' }
      ],
      value: { body: [ { type: 'text', value: 'Foo' } ] }
    },
    {
      keys: [
        { type: '*', value: '2' },
        { type: '*', value: 'bar' }
      ],
      value: { body: [ { type: 'text', value: 'Foo' } ] }
    }
  ]
}
```

I don't think there's a better place in the structure to put this information.

File: spec/data-model.md
Comment (eemeli) [reply spec/data-model.md:78]: This is required to deal with changes that may arise in MessageFormat 2.1 or later, if we do choose to make some use of the reserved characters.

Let's say we add inline comments with a syntax like `{# ... #}`. If this happens, we'll want to also be able to represent these in the data model, where they'll show up in the `Expression` body. This sentence lets us do that so that a MessageFormat 2.0 processor won't change its behaviour if it's fed a MessageFormat 2.1 data model.

File: spec/data-model.md
Comment (eemeli) [reply spec/data-model.md:82]: There is no difference; they're meant to be synonymous.

It might make sense to use "array" throughout, as that's really the JS/TS implementation of a "list".

File: spec/data-model.md
Comment (eemeli) [reply spec/data-model.md:152]: Yes, because currently non-`sigil` characters in this position may produce a parsing error, and that it's arguable that changing a broken string to a not-broken one could be considered a breaking change and require a 3.0 update.

The set should be sufficiently large to cover reasonable future needs.

File: spec/data-model.md
Comment (eemeli) [reply spec/data-model.md:162]: How so?

File: spec/data-model.md
Comment (eemeli) [reply spec/data-model.md:168]: One reasonable extension point could be adding something like `span: Span` to all interface, with a definition like:

```ts
interface Span {
  start: int
  end: int
}
```

Is there a reason to prevent something like this from being done? Or am I misunderstanding your concern here?

File: spec/data-model.md
Comment (eemeli) [reply spec/data-model.md:73]: The way the doc is currently structured, the prose acts as an introduction to the subsequent TS definitions. Separating a "Text" section would change that structure.

Would it help if the "Expressions" section was retitled, say to "Expression contents" or something similar?

File: spec/data-model.md
Comment (eemeli) [reply spec/data-model.md:121]: I'm not immediately convinced of the benefits of defining an atomic/simple vs. compound split, as in practice this would be splitting a set of three things into two different categories, at the cost of introducing novel terminology.

What would we gain from this?

File: spec/message.json
Comment (eemeli) [reply spec/message.json:131]: There's also the approach @stasm has suggested, i.e. expressing single-pattern messages as having an empty list for selectors, and a single entry in variants with a corrspondingly empty list of keys.

As I mentioned on the call, I'd much prefer landing something first and then iterating on it, rather than sorting out the details all in one go.

File: spec/data-model.md
Comment (catamorphism) [reply spec/data-model.md:84]: Yes, I suggest using one or the other throughout, as otherwise it might seem like there's meaning behind the choice of the `Array` notation in some cases and `[]` in others.

File: spec/data-model.md
Comment (catamorphism) [reply spec/data-model.md:121]: I thought it would be better-organized that way, but it's true that it's a relatively small set of things. In the interest of not increasing the amount of work left to do, I'll withdraw this suggestion.

File: spec/message.json
Comment (catamorphism) [reply spec/message.json:25]: Disregard; this goes with the other suggestion that I withdrew.

**Review by catamorphism — COMMENTED** [review]
Looks good. I agree with your intention to land something and iterate it on it more later, so in that spirit, I think the only one of my suggestions that needs to be addressed is the clarification about what the fields of a `Reserved` are.

File: spec/data-model.md
Comment (eemeli) [reply spec/data-model.md:164]: Yes, and those literals are still available within the raw `source`. The `operand` we can extract as it's before the _annotation_, but I don't really see why we should present the _reserved_ itself with any distinct shape. It's not really meant to be processed.

File: spec/data-model.md
Comment (eemeli) [reply spec/data-model.md]: Changed to `name: string`.

File: spec/data-model.md
Comment (eemeli) [reply spec/data-model.md:86]: Changed the one mention of "ordered list" to "array".

File: spec/data-model.md
Comment (eemeli) [reply spec/data-model.md]: Updated as suggested.

File: spec/data-model.md
Comment (catamorphism) [reply spec/data-model.md:166]: I see, thanks for the clarification!

**Review by catamorphism — APPROVED** [review]
Looks great!

File: spec/data-model.md
Comment (stasm) [inline spec/data-model.md:96]: A `FunctionRef` can also have a `Literal` or a `VariableRef` as an argument, and in fact, due to how our syntax is designed, I'd argue that the function's argument is more important than the function. (E.g. it comes first in the syntax.)

I'd like to suggest an alternative way to structure our _expressions_, to more closely map to our syntax:

```abnf
expression = "{" [s] ((operand [s annotation]) / annotation) [s] "}"
```

Let's special-case argument-less functions rather than function-less operands.

Instead of `Literal | VariableRef | FunctionRef` here and `operand?: Literal | VariableRef` inside `FunctionRef`, we can do:

```ts
type Expression = OperandExpr | FunctionExpr;
interface OperandExpr {
    operand: Literal | VariableRef;
    annotation?: FunctionExpr;
}
interface FunctionExpr {
    operand: Literal | VariableRef;
    annotation?: FunctionExpr;
}
```

FWIW, this is how I implemented expressions in [stasm/message2](https://github.com/stasm/message2): 
https://github.com/stasm/message2/blob/4abf43f2023b6e20d8ee1d462684d0741ece791b/syntax/ast.ts#L44-L70

(Not blocking this PR on this, but I'd like to discuss this change as a follow-up.)

**Review by stasm — APPROVED** [review]
As agreed in the yesterday's meeting, we'll land this without the JSOM schema for now.

File: spec/data-model.md
Comment (eemeli) [inline spec/data-model.md:18]: @aphillips @stasm This is the paragraph that'll need to be replaced to drop the JSON schema.

File: spec/data-model.md
Comment (eemeli) [reply spec/data-model.md:96]: I will be happy to discuss this further in a follow-on issue or PR.

**Comment by eemeli** [comment]
To make sure that it's usable for future work after this lands, the JSON schema is still available here:
https://github.com/eemeli/message-format-wg/blob/json-schema-data-model/spec/message.json

**Review by aphillips — APPROVED** [review]

File: spec/data-model.md
Comment (stasm) [reply spec/data-model.md:94]: Filed #426 to continue this.

File: spec/message.json
Comment (stasm) [reply spec/message.json]: I filed #437 to continue discussing this.

**Pull request merged**
